test/e2e/consumer_credit_flow_policy.test.ts

```diff
@@ -0,0 +1,170 @@
+    { content: Buffer.from("hello") },
+    { content: Buffer.from("there") },
+    { content: Buffer.from("brave") },
+    { content: Buffer.from("new") },
+    { content: Buffer.from("world") },
+  ]
+  const chunks = [chunk, chunk]
+  const nMessages = chunk.length * chunks.length
+
+  before(() => {
+    process.env.MAX_SHARED_CLIENT_INSTANCES = "10"
+  })
+
+  after(() => {
+    if (previousMaxSharedClientInstances !== undefined) {
+      process.env.MAX_SHARED_CLIENT_INSTANCES = previousMaxSharedClientInstances
+      return
+    }
+    delete process.env.MAX_SHARED_CLIENT_INSTANCES
+  })
+
+  beforeEach(async () => {
+    client = await createClient(username, password)
+    streamName = createStreamName()
+    await rabbit.createStream(streamName)
+    publisher = await createPublisher(streamName, client)
+  })
+
+  afterEach(async () => {
+    try {
+      sandbox.restore()
+      await client.close()
+      await rabbit.deleteStream(streamName)
+      await rabbit.closeAllConnections()
+      await rabbit.deleteAllQueues({ match: /my-stream-/ })
+    } catch (_e) {}
+  })
+
+  it("NewCreditOnChunkReceived policy requests new credit when chunk is completely received", async () => {
+    const consumerRef = createConsumerRef()
+    const policy = creditsOnChunkReceived(1, 1)
+    sandbox.on(policy, "onChunkReceived")
+    const received: Message[] = []
+    await client.declareConsumer(
+      {
+        stream: streamName,
+        offset: Offset.first(),
+        singleActive: true,
+        consumerRef: consumerRef,
+        creditPolicy: policy,
+      },
+      (message: Message) => {
+        received.push(message)
+      }
+    )
+
+    await send(publisher, chunks)
+
+    await eventually(() => expect(received.length).eql(nMessages))
+    await eventually(() => expect(policy.onChunkReceived).called.exactly(chunks.length))
+    await always(() => expect(policy.onChunkReceived).called.below(chunks.length + 1), 5000)
+  }).timeout(10000)
+
+  it("NewCreditOnChunkProgress policy requests new credit when chunk handled at 50% progress", async () => {
+    const consumerRef = createConsumerRef()
+    const policy = creditsOnChunkProgress(1, 0.5, 1)
+    sandbox.on(policy, "onChunkProgress")
+    const received: Message[] = []
+    await client.declareConsumer(
+      {
+        stream: streamName,
+        offset: Offset.first(),
+        singleActive: true,
+        consumerRef: consumerRef,
+        creditPolicy: policy,
+      },
+      (message: Message) => {
+        received.push(message)
+      }
+    )
+
+    await send(publisher, chunks)
+
+    await eventually(() => expect(received.length).eql(nMessages))
+    await eventually(() => expect(policy.onChunkProgress).called.exactly(nMessages))
+    await always(() => expect(policy.onChunkProgress).called.below(nMessages + 1), 5000)
+  }).timeout(10000)
+
+  it("NewCreditOnChunkProgress policy requests new credit when chunk handled at 100% progress", async () => {
```

Comment:
Maybe we can measure the timing by having the consumer callback set a flag if the `received` becomes bigger than the ratio*length before the spy has been called? Basically, measure that, if ratio is 0.5, we don't have 0.6*length messages handled before the spy has been called, and if so set a flag we can expect to.